### PR TITLE
Reject references to self in super constructor calls

### LINKED ISSUE #11208: VerifyError with implicit ClassTag in companion object
The following throws a `VerifyError` (Tested with 3.0.0-M3)
```
import scala.reflect.ClassTag

@main def run = println(Foo)

abstract class Bar[T](implicit val thisClassTag: ClassTag[T])

class Foo
object Foo extends Bar[Foo]
```

Alternatives tried:
- Changing `abstract class Bar[T](implicit val thisClassTag: ClassTag[T])` to `abstract class Bar[T: ClassTag]{ val thisClassTag: ClassTag[T] = implicitly[ClassTag[T]] }` works
- Changing `object Foo extends Bar[Foo]` to `object Foo extends Bar()(classTag[Foo])` works
- Summoning an implicit of a different class works (only `ClassTag` specifically seems to be affected)
- Using the ClassTag of anything other than the companion class works (`object Foo extends Bar[Foo2]` seems fine)

### LINKED ISSUE #12557: java.lang.VerifyError: Bad type on operand stack
## Compiler version

`3.0.1-RC1-bin-20210521-de319fb-NIGHTLY` and `3.0.0`

## Minimized code

```scala
package example

abstract class X[P <: Product](using val m: scala.deriving.Mirror.ProductOf[P]) {
  def unapply(p: P): m.MirroredElemTypes = ???
}

case class A(a: Int)
object A extends X[A]

object Main {
  def main(args: Array[String]): Unit = {
    A.unapply(A(2))
  }
}
```

## Output

```
[info] running example.Main 
[error] (run-main-7) java.lang.VerifyError: Bad type on operand stack
[error] Exception Details:
[error]   Location:
[error]     example/A$.<init>()V @2: invokevirtual
[error]   Reason:
[error]     Type uninitializedThis (current frame, stack[1]) is not assignable to 'example/A$'
[error]   Current Frame:
[error]     bci: @2
[error]     flags: { flagThisUninit }
[error]     locals: { uninitializedThis }
[error]     stack: { uninitializedThis, uninitializedThis }
[error]   Bytecode:
[error]     0x0000000: 2a2a b600 15b7 0018 b1                 
[error] java.lang.VerifyError: Bad type on operand stack
[error] Exception Details:
[error]   Location:
[error]     example/A$.<init>()V @2: invokevirtual
[error]   Reason:
[error]     Type uninitializedThis (current frame, stack[1]) is not assignable to 'example/A$'
[error]   Current Frame:
[error]     bci: @2
[error]     flags: { flagThisUninit }
[error]     locals: { uninitializedThis }
[error]     stack: { uninitializedThis, uninitializedThis }
[error]   Bytecode:
[error]     0x0000000: 2a2a b600 15b7 0018 b1                 
[error] 
[error] 	at example.Main$.main(A.scala:11)
[error] 	at example.Main.main(A.scala)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
```

### PR DESCRIPTION
**Pull request by odersky**
Fixes #11208
Fixes #12557

### CODE REVIEW

**Comment by griggt** [comment]
Looks like there is a typo in the issue reference (and also in the file names of the associated tests), it should be #11208.

**Comment by odersky** [comment]
@griggt I fixed the issue reference, the tests have the right names already.

**Comment by griggt** [comment]
@odersky the issue reference and test names are consistent now, but the digits are still transposed: 11028 vs. 11208 (correct).

**Comment by odersky** [comment]
@griggt Thanks for pointing it out!

File: compiler/src/dotty/tools/dotc/transform/CheckNoSuperThis.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  import tpd._
+
+  override def phaseName: String = CheckNoSuperThis.name
+
+  override def runsAfterGroupsOf: Set[String] = Set(Constructors.name)
+
+  override def transformDefDef(mdef: DefDef)(using Context): DefDef =
+    if mdef.symbol.isClassConstructor then
+      mdef.rhs match
+        case Block(stats, _) => splitAtSuper(stats) match
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/CheckNoSuperThis.scala:28]: Constructors#transformTemplate also calls splitAtSuper on the constructor, any particular reason the checks here couldn't be done there?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/CheckNoSuperThis.scala:28]: We also need to check secondary constructors. It seems to be simpler to do this in a separate miniphase.

**Review by smarter — APPROVED** [review]

**Pull request merged**